Add meta support to tensor range factories

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #67032

This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

cbalioglu committed:
Add meta support to tensor range factories

This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Add meta support to tensor range factories"


This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Add meta support to tensor range factories"


This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

[ghstack-poisoned]

cbalioglu commented on test/test_tensor_creation_ops.py:
```diff
@@ -3170,6 +3171,58 @@ def test_tensor_from_non_writable_numpy(self, device):
             self.assertEqual(len(w), 0)
 
 
+class TestTensorCreationMeta(TestCase):
```
@bdhirsh any feedback on my last comment?

bdhirsh replied:
Hey, sorry for the delay.

> internally calls torch.isnan which does not support the meta backend. So assertEqual itself raises NotImplementedError and the whole test gets "swallowed" before any assertions could take place. 

Oh okay, so most of the existing non-meta tests don't really test meta tensors well, because of this?

I think two reasonable options are problably either:

(1) Just get `isnan` to work properly with meta tensors, so you can re-use those tests

(2) Keep the meta-specific `TestTensorCreationMeta` tests you have around for now, but with a big comment that says they should be fully killed once `isnan` works with meta tensors.

I think (1) would be better :) But I totally get it if you don't want to add scope to what you're working on

cbalioglu committed:
Update on "Add meta support to tensor range factories"


This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Add meta support to tensor range factories"


This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Add meta support to tensor range factories"


This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

[ghstack-poisoned]

cbalioglu committed:
Update on "Add meta support to tensor range factories"


This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

[ghstack-poisoned]

bdhirsh commented:
> @bdhirsh in my latest revision I "fixed" `assertEqual()` by shortcutting it for meta tensors. Unfortunately it revealed quite a few unit tests that were supposed to fail, but were being skipped due to `assertEqual()` raising a `NotImplementedError`. Since I am not familiar with those tests and what they test, I prefer not to touch them. Although not ideal, let me know if you are okay with me merging this PR without any specific tests (just the changes in `aten/`); otherwise, I will postpone this PR to a future date after the quirks of device-specific tests are resolved.
> 
> Also putting a reference to #67190 here for additional context.

Okay, I think that's fine. Can you add a comment above them, mentioning that they should go away (and link to your comment about why we can't do that for now?)

bdhirsh reviewed (approved):
LGTM (but add the comments described above please)

cbalioglu committed:
Update on "Add meta support to tensor range factories"


This PR adds meta backend support to the `range`, `arange`, `linspace`, and `logspace` operators.

Note that the original PR (#66630) was reverted due to two failing unit tests in the Bionic CI. This revision includes a fix for those tests; otherwise its content is identical to the previous PR.

Original commit changeset: 2f9d8d1acbb0

Differential Revision: [D31834403](https://our.internmc.facebook.com/intern/diff/D31834403/)

[ghstack-poisoned]
